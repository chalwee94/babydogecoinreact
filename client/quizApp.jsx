import * as React from "react";
import { useContext, useEffect, useState } from "react";
import { Routes, Route, Link } from "react-router-dom";
import { isCorrectAnswer, randomQuestion } from "../server/questions";

export const QuestionContext = React.createContext({ randomQuestion });

export function FrontPage() {
  const [question, setQuestion] = useState();
  useEffect(async () => {
    const res = await fetch("/api/question");
    setQuestion(await res.json());
  }, []);

  return (
    <div>
      <h1>Quiz</h1>
      <Link to={"/question"}>
        <button>Show me a random question</button>
      </Link>
    </div>
  );
}

export function ShowQuestion() {
  const { randomQuestion } = useContext(QuestionContext);
  const [question, setQuestion] = useState(randomQuestion());
  const [answer, setAnswer] = useState();

  function handleSubmit(answer) {
    setQuestion(randomQuestion());
    const setAnswerStatus = setAnswer;

    if (isCorrectAnswer(question, answer)) {
      console.log("right");
      return setAnswerStatus("Correct");
    } else {
      console.log("wrong");
      setAnswerStatus("Incorrect");
    }
  }

  return (
    <>
      <h3>{question.question}</h3>
      {Object.keys(question.answers)
        .filter((a) => question.answers[a])
        .map((a) => (
          <div key={a} data-testid={a}>
            <button onClick={() => handleSubmit(a)}>
              {question.answers[a]}
            </button>
          </div>
        ))}
      <div>
        <h3>Your answer is: </h3>
        <h4>{answer || ""}</h4>
      </div>
    </>
  );
}

export function Quiz({ question, setAnswer }) {
  return (
    <Routes>
      <Route path={"/"} element={<FrontPage />} />
      <Route
        path={"/question"}
        element={<ShowQuestion question={question} answer={setAnswer} />}
      />
    </Routes>
  );
}
