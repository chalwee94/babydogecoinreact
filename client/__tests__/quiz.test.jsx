import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import { ShowAnswer, ShowQuestion, QuestionContext } from "../quizApp";
import { Simulate } from "react-dom/test-utils";

const question = {
  question: "Do you like Jest?",
  answers: {
    answer_a: "Yes",
    answer_b: "No",
    answer_c: "Jest sucks",
  },
  correct_answers: {
    answer_a_correct: "false",
    answer_b_correct: "false",
    answer_c_correct: "true",
  },
};

describe("Quiz", () => {
  it("shows random question", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/question"]}>
        <QuestionContext.Provider value={{ randomQuestion: () => question }}>
          <ShowQuestion />
        </QuestionContext.Provider>
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
  });

  /*
  it("shows random question", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/question"]}>
        <QuestionContext.Provider value={{ randomQuestion: () => question }}>
          <ShowQuestion />
        </QuestionContext.Provider>
      </MemoryRouter>,
      expect(element.innerHTML).toMatchSnapshot()
    );
  });

   */

  it("shows answer", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/question"]}>
        <QuestionContext.Provider value={{ randomQuestion: () => question }}>
          <ShowQuestion />
        </QuestionContext.Provider>
      </MemoryRouter>,
      element
    );

    Simulate.click(element.querySelector("[data-testid=answer_a] button"));
    //expect(setAnswer).toBeCalled();
    expect(element.innerHTML).toMatchSnapshot();
  });
});
