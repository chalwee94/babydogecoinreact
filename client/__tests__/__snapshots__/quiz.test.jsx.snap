// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz shows answer 1`] = `"<h3>Do you like Jest?</h3><div data-testid=\\"answer_a\\"><button>Yes</button></div><div data-testid=\\"answer_b\\"><button>No</button></div><div data-testid=\\"answer_c\\"><button>Jest sucks</button></div><div><h3>Your answer is: </h3><h4>Incorrect</h4></div>"`;

exports[`Quiz shows random question 1`] = `"<h3>Do you like Jest?</h3><div data-testid=\\"answer_a\\"><button>Yes</button></div><div data-testid=\\"answer_b\\"><button>No</button></div><div data-testid=\\"answer_c\\"><button>Jest sucks</button></div><div><h3>Your answer is: </h3><h4></h4></div>"`;
